src/Microsoft.AspNet.Mvc.Core/ModelBinding/GenericModelBinder.cs

```diff
@@ -13,83 +13,129 @@ public class GenericModelBinder : IModelBinder
         {
-            return GetGenericBinderType(
-                        typeof(IDictionary<,>),
-                        typeof(Dictionary<,>),
-                        typeof(DictionaryModelBinder<,>),
-                        modelType);
+            return GetGenericModelBindingInfo(
+                typeof(IDictionary<,>),
+                typeof(Dictionary<,>),
+                typeof(DictionaryModelBinder<,>),
+                modelType);
         }
 
-        private static Type GetKeyValuePairBinder(Type modelType)
+        private static GenericModelBindingInfo GetKeyValuePairBinderInfo(Type modelType)
         {
             var modelTypeInfo = modelType.GetTypeInfo();
             if (modelTypeInfo.IsGenericType &&
                 modelTypeInfo.GetGenericTypeDefinition() == typeof(KeyValuePair<,>))
             {
-                return typeof(KeyValuePairModelBinder<,>).MakeGenericType(modelTypeInfo.GenericTypeArguments);
+                var modelBinderType = typeof(KeyValuePairModelBinder<,>)
+                    .MakeGenericType(modelTypeInfo.GenericTypeArguments);
+
+                return new GenericModelBindingInfo()
+                {
+                    ModelBinderType = modelBinderType,
+                    UnderlyingModelType = modelType,
+                };
             }
 
             return null;
         }
 
-        /// <remarks>
-        /// Example: <c>GetGenericBinderType(typeof(IList&lt;T&gt;), typeof(List&lt;T&gt;),
-        /// typeof(ListBinder&lt;T&gt;), ...)</c> means that the <c>ListBinder&lt;T&gt;</c> type can update models that
-        /// implement <see cref="IList{T}"/>, and if for some reason the existing model instance is not updatable the
-        /// binder will create a <see cref="List{T}"/> object and bind to that instead. This method will return
-        /// <c>ListBinder&lt;T&gt;</c> or <c>null</c>, depending on whether the type and updatability checks succeed.
-        /// </remarks>
-        private static Type GetGenericBinderType(Type supportedInterfaceType,
-                                                 Type newInstanceType,
-                                                 Type openBinderType,
-                                                 Type modelType)
+        //
+        // Example: 
+        //      GetGenericBinderType(typeof(IList<T>), typeof(List<T>), typeof(ListBinder<T&>), ...)
+        //
+        // This means that the ListBinder<T> type can work with models that implement IList<T>, and if there is no 
+        // existing model instance, the binder will create a List{T}. 
+        //
+        // This method will return null if the given model type isn't compatible with the combination of
+        // supportedInterfaceType and modelType. If supportedInterfaceType and modelType are compatible, then
+        // it will return the closed-generic newInstanceType and closed-generic openBinderType.
```

Comment:
"closed generic" (In addition to deleting things, you also like hyphenating things that aren't hyphenated terms.)